### PR TITLE
crimson/osd: cleanups to drop some do_with() calls

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test api

File: src/crimson/osd/replicated_recovery_backend.cc [hunk 1]
```diff
@@ -745,29 +719,29 @@ seastar::future<> ReplicatedRecoveryBackend::_handle_push(
...
-    return submit_push_data(pop.recovery_info, first, complete, clear_omap,
-	data_zeros, pop.data_included, data, pop.omap_header, pop.attrset,
-	pop.omap_entries, t).then([this, complete, &pop, t] {
-      if (complete) {
-	pg.get_recovery_handler()->on_local_recover(pop.recovery_info.soid,
-			    pop.recovery_info, false, *t);
-      }
-    });
+  return submit_push_data(pop.recovery_info, first, complete, clear_omap,
+        data_zeros, pop.data_included, pop.data, pop.omap_header,
```
Comment (xxhdx1985126) [inline src/crimson/osd/replicated_recovery_backend.cc:738]: data_zeros may be captured by reference in [submit_push_data](https://github.com/ceph/ceph/blob/7ef6453e4aff9bcef8bb2dd26182dd35bb98dee0/src/crimson/osd/replicated_recovery_backend.cc#L989), I think that may also need to be modified

Comment (tchaikov) [reply src/crimson/osd/replicated_recovery_backend.cc]: @xxhdx1985126 fixed and repushed.

**Comment by tchaikov** [comment]
jenkins test make check

**Review by rzarzynski — APPROVED** [review]

File: src/crimson/osd/replicated_recovery_backend.cc [hunk 2]
```diff
@@ -27,7 +27,24 @@ seastar::future<> ReplicatedRecoveryBackend::recover_object(
   assert(is_recovering(soid));
   // start tracking the recovery of soid
   return maybe_pull_missing_obj(soid, need).then([this, soid, need] {
-    return maybe_push_shards(soid, need);
+    auto& recovery_waiter = recovering.at(soid);
+    if (recovery_waiter.obc) {
```
Comment (xxhdx1985126) [inline src/crimson/osd/replicated_recovery_backend.cc:31]: this condition check is problematic, I've just submitted a [PR](https://github.com/ceph/ceph/pull/38646) for this, please take a look:-)

**Pull request merged**